contracts/roles.js

```diff
@@ -0,0 +1,917 @@
+  const upRole = role;
+  if (upRole.tokenBalances) {
+    const tIndex = upRole.tokenBalances.findIndex(t => t.symbol === token.symbol);
+    if (tIndex === -1) {
+      upRole.tokenBalances.push({ symbol: token.symbol, quantity });
+    } else {
+      upRole.tokenBalances[tIndex].quantity = api.BigNumber(upRole.tokenBalances[tIndex].quantity)
+        .plus(quantity)
+        .toFixed(token.precision, api.BigNumber.ROUND_DOWN);
+    }
+  } else {
+    upRole.tokenBalances = [
+      { symbol: token.symbol, quantity },
+    ];
+  }
+  await api.db.update('roles', upRole);
+}
+
+actions.deposit = async (payload) => {
+  const {
+    roleId, symbol, quantity,
+    isSignedWithActiveKey,
+  } = payload;
+
+  const depToken = await api.db.findOneInTable('tokens', 'tokens', { symbol });
+  if (!api.assert(isSignedWithActiveKey === true, 'you must use a custom_json signed with your active key')
+    || !api.assert(typeof roleId === 'string' && api.BigNumber(roleId).isInteger(), 'invalid roleId')
+    || !api.assert(typeof quantity === 'string' && api.BigNumber(quantity).gt(0), 'invalid quantity')
+    || !api.assert(api.BigNumber(quantity).dp() <= depToken.precision, 'quantity precision mismatch')) {
+    return;
+  }
+
+  const role = await api.db.findOne('roles', { _id: api.BigNumber(roleId).toNumber() });
+  if (api.assert(role, 'role not found') && api.assert(role.active, 'role must be active to deposit')) {
+    const res = await api.executeSmartContract('tokens', 'transferToContract', { symbol, quantity, to: ContractName });
+    if (res.errors === undefined
+      && res.events && res.events.find(el => el.contract === 'tokens' && el.event === 'transferToContract' && el.data.from === api.sender && el.data.to === ContractName && el.data.quantity === quantity) !== undefined) {
+      await updateTokenBalances(role, depToken, quantity);
+      api.emit('deposit', {
+        roleId,
+        symbol,
+        quantity,
+      });
+    }
+  }
+};
+
+actions.receiveDtfTokens = async (payload) => {
+  const {
+    data, symbol, quantity,
+    callingContractInfo,
+  } = payload;
+
+  if (!api.assert(callingContractInfo && callingContractInfo.name === 'tokenfunds', 'not authorized')) return;
+  if (!api.assert(typeof data === 'object'
+    && data.constructor.name === 'Object'
+    && 'roleId' in data && typeof data.roleId === 'string'
+    && api.BigNumber(data.roleId).isInteger(), 'invalid incoming payload')) return;
+
+  const role = await api.db.findOne('roles', { _id: api.BigNumber(data.roleId).toNumber() });
+  if (api.assert(role, 'role not found') && api.assert(role.active, 'role must be active to deposit')) {
+    const depToken = await api.db.findOneInTable('tokens', 'tokens', { symbol });
+    await updateTokenBalances(role, depToken, quantity);
+    api.emit('receiveDtfTokens', {
+      roleId: data.roleId,
+      symbol,
+      quantity,
+    });
+  }
+};
+
+actions.receiveDistTokens = async (payload) => {
```

Comment:
The code for this function is almost exactly the same as for receiveDtfTokens.  Perhaps you could introduce a helper function used by both of these, to avoid so much code duplication?